route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,210 @@ namespace route_following_plugin
-                    end_dist  = wm_->routeTrackPos(shortest_path[last_lanelet_index + 1].centerline2d().back()).downtrack;
-                }               
-
-                lane_change_start_dist = current_progress;
-                double starting_lanelet_id = shortest_path[last_lanelet_index].id();
-                double ending_lanelet_id = shortest_path[last_lanelet_index+1].id();
-                resp.new_plan.maneuvers.push_back(
-                composeLaneChangeManeuverMessage(lane_change_start_dist, end_dist, speed_progress, target_speed, 
-                                    starting_lanelet_id, ending_lanelet_id,
-                                    time_progress));
-                
-                ++last_lanelet_index;
+            else{
+                maneuvers.push_back(composeLaneFollowingManeuverMessage(start_dist, end_dist, prev_mvr_end_speed, target_speed_in_lanelet, route_shortest_path[i].id(),ros::Time::now()));
             }
-            else
-            {
-                resp.new_plan.maneuvers.push_back(
-                composeManeuverMessage(current_progress, end_dist,  
-                                    speed_progress, target_speed, 
-                                    shortest_path[last_lanelet_index].id(), time_progress));
-                
-            }
-            current_progress += dist_diff;
-            //time_progress = resp.new_plan.maneuvers.back().lane_following_maneuver.end_time;
-            speed_progress = target_speed;
-            
+            prev_mvr_end_speed = target_speed_in_lanelet;
+        }
+        //add lane follow as last maneuver
+        if(maneuvers.size() < route_shortest_path.size()){
+            double target_speed_in_lanelet = findSpeedLimit(route_shortest_path.back());
+            double start_dist = wm_->routeTrackPos(route_shortest_path.back().centerline2d().front()).downtrack;
+            double end_dist = wm_->routeTrackPos(route_shortest_path.back().centerline2d().back()).downtrack; 
+            maneuvers.push_back(composeLaneFollowingManeuverMessage(start_dist, end_dist, prev_mvr_end_speed, target_speed_in_lanelet, route_shortest_path.back().id(),ros::Time::now()));
+        }
+        
+        return maneuvers;
+    }
 
-            if(current_progress >= total_maneuver_length || last_lanelet_index == shortest_path.size() - 1)
-            {
-                break;
-            }
+    bool RouteFollowingPlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp)
+    { 
+        if(latest_maneuver_plan_.empty()){
+            ROS_ERROR_STREAM("A maneuver plan has not been generated");
+            return true;
+        }
 
-            ++last_lanelet_index;
+        lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
+        
+        double current_downtrack = wm_->routeTrackPos(current_loc).downtrack;
+        double speed_progress = current_speed_;
```

Comment:
Neither of these variables is used in this method.  Remove this line.

Reply:
speed_progress is now used (in call to updateStartingSpeed(), but there's no need for it.  Just use current_speed_ in that call, and this local variable can be eliminated.